### ISSUE TITLE
Selecting functions with NNs

### ISSUE THREAD

**Issue by lorepieri8** [lorepieri8/ARC_challenge#23]
I think that looping over functions (like in generate_functions) is here to stay. But I think that this loop should be over few functions, without the necessity for a similarity score, which is misleading whenever an intermediate step is not closer to the final solution.

A way to make this initial set of functions small could be to:

-For every task of in-out pairs extract all the attributes
-Generate a dateset of composite programs (f1(f2(f3(...))) and make them act on the tasks. The final data in the dataset is : task-attributes --> program --> task-attributes_out -->success? (Y/N)
This part can actually be done in multiple ways.
-Make a NNs learn how likely is for a program to solve a task, given specific task-attributes.
-Pick the top predictions and use them as input of generate_functions.

Basically since:
-we know the attributes of t_in
-we suppose to be able to guess the attributes of t_out
this NN would tell us the most likely functions too appear in the program solution.

References:
https://arxiv.org/pdf/1611.01989.pdf
